google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/DdlBatch.java

```diff
@@ -204,36 +212,48 @@ public ApiFuture<Void> writeAsync(CallType callType, Iterable<Mutation> mutation
   public ApiFuture<long[]> runBatchAsync(CallType callType) {
     ConnectionPreconditions.checkState(
         state == UnitOfWorkState.STARTED, "The batch is no longer active and cannot be ran");
-    if (statements.isEmpty()) {
-      this.state = UnitOfWorkState.RAN;
-      return ApiFutures.immediateFuture(new long[0]);
-    }
-    // create a statement that can be passed in to the execute method
-    Callable<long[]> callable =
-        () -> {
-          try {
-            OperationFuture<Void, UpdateDatabaseDdlMetadata> operation =
-                ddlClient.executeDdl(statements);
+    try (Scope ignore = span.makeCurrent()) {
+      if (statements.isEmpty()) {
+        this.state = UnitOfWorkState.RAN;
+        return ApiFutures.transform(
+            asyncEndUnitOfWorkSpan(), unused -> new long[0], MoreExecutors.directExecutor());
+      }
+      // Set the DDL statements on the span.
+      span.setAllAttributes(Attributes.of(DB_STATEMENT_ARRAY_KEY, statements));
+      // create a statement that can be passed in to the execute method
+      Callable<long[]> callable =
+          () -> {
             try {
-              // Wait until the operation has finished.
-              getWithStatementTimeout(operation, RUN_BATCH_STATEMENT);
-              long[] updateCounts = new long[statements.size()];
-              Arrays.fill(updateCounts, 1L);
-              state = UnitOfWorkState.RAN;
-              return updateCounts;
-            } catch (SpannerException e) {
-              long[] updateCounts = extractUpdateCounts(operation);
-              throw SpannerExceptionFactory.newSpannerBatchUpdateException(
-                  e.getErrorCode(), e.getMessage(), updateCounts);
+              OperationFuture<Void, UpdateDatabaseDdlMetadata> operation =
+                  ddlClient.executeDdl(statements);
+              try {
+                // Wait until the operation has finished.
+                getWithStatementTimeout(operation, RUN_BATCH_STATEMENT);
+                long[] updateCounts = new long[statements.size()];
+                Arrays.fill(updateCounts, 1L);
+                state = UnitOfWorkState.RAN;
+                return updateCounts;
+              } catch (SpannerException e) {
+                long[] updateCounts = extractUpdateCounts(operation);
+                throw SpannerExceptionFactory.newSpannerBatchUpdateException(
+                    e.getErrorCode(), e.getMessage(), updateCounts);
+              }
+            } catch (Throwable t) {
+              state = UnitOfWorkState.RUN_FAILED;
```

Comment:
Do we want to capture span status here as failed ?

Reply:
Good catch, done.